source/common/upstream/outlier_detection_impl.h

```diff
@@ -130,19 +159,45 @@ class DetectorHostMonitorImpl : public DetectorHostMonitor {
   const absl::optional<MonotonicTime>& lastUnejectionTime() override {
     return last_unejection_time_;
   }
-  double successRate() const override { return success_rate_; }
+
+  // Template based and function based methods for managing SuccessRate detectors
+  template <SuccessRateMonitorType Type>
+  const std::unique_ptr<SuccessRateMonitor>& getSRMonitor() const {
+    return success_rate_monitors_.at(Type);
+  }
+  const std::unique_ptr<SuccessRateMonitor>& getSRMonitor(SuccessRateMonitorType Type) const {
+    return success_rate_monitors_.at(Type);
+  }
+
+  static EjectionType getEjectionType(SuccessRateMonitorType type);
+
+  double successRate(SuccessRateMonitorType type) const override {
+    return getSRMonitor(type)->getSuccessRate();
+  }
+  void updateCurrentSuccessRateBucket();
+  void successRate(SuccessRateMonitorType type, double new_success_rate) {
+    getSRMonitor(type)->setSuccessRate(new_success_rate);
+  }
+
+  // handlers for reporting local origin errors
+  void localOriginFailure();
+  void localOriginNoFailure();
 
 private:
   std::weak_ptr<DetectorImpl> detector_;
   std::weak_ptr<Host> host_;
-  std::atomic<uint32_t> consecutive_5xx_{0};
-  std::atomic<uint32_t> consecutive_gateway_failure_{0};
   absl::optional<MonotonicTime> last_ejection_time_;
   absl::optional<MonotonicTime> last_unejection_time_;
   uint32_t num_ejections_{};
-  SuccessRateAccumulator success_rate_accumulator_;
-  std::atomic<SuccessRateAccumulatorBucket*> success_rate_accumulator_bucket_;
-  double success_rate_;
+
+  // counters for externally generated failures
+  std::atomic<uint32_t> consecutive_5xx_{0};
+  std::atomic<uint32_t> consecutive_gateway_failure_{0};
+
+  // counters for local origin failures
+  std::atomic<uint32_t> consecutive_local_origin_failure_{0};
+
+  std::map<SuccessRateMonitorType, std::unique_ptr<SuccessRateMonitor>> success_rate_monitors_;
```

Comment:
Did you consider `absl::flat_hash_map` for this?

Reply:
Thanks for suggestion. Converted std::map to absl::flat_map_hash.